hare3/runner/runner.go

```diff
@@ -0,0 +1,394 @@
+	"github.com/spacemeshos/go-spacemesh/datastore"
+	"github.com/spacemeshos/go-spacemesh/hare"
+	heligibility "github.com/spacemeshos/go-spacemesh/hare/eligibility"
+	"github.com/spacemeshos/go-spacemesh/hare3"
+	"github.com/spacemeshos/go-spacemesh/hare3/broker"
+	"github.com/spacemeshos/go-spacemesh/hare3/eligibility"
+	"github.com/spacemeshos/go-spacemesh/hare3/leader"
+	"github.com/spacemeshos/go-spacemesh/hare3/weakcoin"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/p2p/pubsub"
+	"github.com/spacemeshos/go-spacemesh/signing"
+	"github.com/spacemeshos/go-spacemesh/syncer"
+)
+
+// WeakCoinResult is used to communicate the result of the weak coin back to
+// the application, via a channel.
+type WeakCoinResult struct {
+	Layer types.LayerID
+	// A nil coin value means that no coins were received.
+	Coin *bool
+}
+
+// Protocol runner runs an instance of the hare3 protocol.
+type ProtocolRunner struct {
+	clock        RoundClock
+	protocol     hare3.Protocol
+	coinChooser  *weakcoin.Chooser
+	maxRound     hare3.AbsRound
+	gossiper     NetworkGossiper
+	layer        types.LayerID
+	ac           *eligibility.ActiveCheck
+	mb           *MessageBuilder
+	weakcoinChan chan<- WeakCoinResult
+	l            log.Log
+}
+
+func NewProtocolRunner(
+	clock RoundClock,
+	protocol *hare3.Protocol,
+	coinChooser *weakcoin.Chooser,
+	iterationLimit int8,
+	gossiper NetworkGossiper,
+	layer types.LayerID,
+	ac *eligibility.ActiveCheck,
+	mb *MessageBuilder,
+	weakcoinChan chan<- WeakCoinResult,
+	l log.Log,
+) *ProtocolRunner {
+	return &ProtocolRunner{
+		clock:        clock,
+		protocol:     *protocol,
+		coinChooser:  coinChooser,
+		maxRound:     hare3.NewAbsRound(iterationLimit, 0),
+		gossiper:     gossiper,
+		layer:        layer,
+		ac:           ac,
+		mb:           mb,
+		weakcoinChan: weakcoinChan,
+		l:            l,
+	}
+}
+
+// Run runs the protocol until it terminates and returns the result. If the
+// protocol exceeded the iteration limit or the context expires an error will be
+// returned and the result will be nil.
+func (r *ProtocolRunner) Run(ctx context.Context) ([]types.Hash20, error) {
+	// ok want to actually use a lock here
+	for {
+		round := r.protocol.Round()
+		if round == r.maxRound {
+			return nil, fmt.Errorf("hare protocol runner exceeded iteration limit of %d", r.maxRound.Iteration())
+		}
+		// The weak coin is calculated from pre-round messages, we select the
+		// coin after one round, we are not concerned about late messages
+		// affecting the outcome since the coin is weak (see package doc for
+		// hare3/weakcoin for an explanation of weak)
+		if round == hare3.Preround.Round()+1 {
+			r.weakcoinChan <- WeakCoinResult{
+				Layer: r.layer,
+				Coin:  r.coinChooser.Choose(),
+			}
+		}
+		select {
+		// We await the beginning of the round, which is achieved by calling AwaitEndOfRound with (round - 1).
+		case <-r.clock.AwaitEndOfRound(uint32(round - 1)):
```

Comment:
this is another part where i am confused, based on my understanding protocol needs to wait for 5 network delays (in iteration 0, and then 4). according to this clock it looks like it will wait for soft and hard llock too

- preround -> WAIT -> propose (only in first)
- propose -> WAIT -> WAIT -> commit
- commit -> WAIT -> notify
- notify -> WAIT -> TERMINATED ?

Reply:
If you're referring to the note mentioning that round 1 (soft-lock) is not required in iteration 0,  I understood that as meaning it's not required to execute the code of round 1 as opposed to meaning that you go straight from round 0 to round 2 with only one WAIT.

The reason for my interpretation is that otherwise the specification of how valid values is calculated does not make sense, since they are intended to be calculated at the beginning of rounds 0, 1, 2 and 3.

Reply:
no, this is not what i was referring to. this code will run waits in the wrong places